src/dotty/tools/dotc/core/TypeApplications.scala

```diff
@@ -177,10 +179,31 @@ object TypeApplications {
     def apply(tp: Type, args: List[Type])(implicit ctx: Context): Type = tp.appliedTo(args)
 
     def unapply(tp: Type)(implicit ctx: Context): Option[(Type, List[Type])] = tp match {
-      case TypeRef(prefix, tpnme.hkApply) => unapp(prefix)
+      case tp: RefinedType if Config.newHK =>
+        var refinements: List[RefinedType] = Nil
+        var tycon = tp.stripTypeVar
+        while (tp.isInstanceOf[RefinedType]) {
```

Comment:
I assume this should be `tycon.isInstanceOf[...]` otherwise you get an infinite loop